WebVersion

caiiiycuk opened the pull request:
Hi,
I prepared a PR to merge changes I made to the WebAssembly version. This PR contains 13 commits, and you can inspect them one by one. They are logically structured, which will make them easier to understand. I started my work on version 1.5.7, and it works perfectly. This PR is a rebase of the changes made for version 1.5.7 over the development branch.

I tested the game, and it works without problems, but it is very slow, especially when loading scenes compared to the 1.5.7 branch. Another regression is the size: the version 1.5.7 build is 3.6 MB, but version 1.6.3 is 8.6 MB, which is a bit strange. I hope to fix these problems in future PRs. I believe merging this is safe because it is a fully working version.

You can find a guide on how to build it in Readme.emscripten.md. Actually, it is just a CMake script, but because of an Emscripten bug and some assumptions, you need to follow a specific order (which can also be improved in the future).

The Emscripten version can only be built on a Linux host (this can be improved), but you can also build a subset of the HTML5 game natively by using the ENABLE_HTML5_BUILD option. This is very handy for debugging web-related problems.

Laserlicht commented:
Already two hints:
- `docs/Readme.emscripten.md` should moved to `docs/developers/Building_WebAssembly.md`
- Shouldn't we perhaps add a GitHub action for a build? To prevent breaking the build with future changes

IvanSavenko commented on CMakeLists.txt:
```diff
@@ -57,6 +58,136 @@ option(ENABLE_CCACHE "Speed up recompilation by caching previous compilations" O
+			BUILD_COMMAND "${CMAKE_MAKE_PROGRAM}" "tbb" "tbbmalloc"
+			INSTALL_COMMAND "${CMAKE_MAKE_PROGRAM}" "install" "tbb" "tbbmalloc"
+	)
+	ExternalProject_Add(vcmi-extras
+			PREFIX "vcmi-extras"
+			GIT_REPOSITORY "https://github.com/vcmi-mods/vcmi-extras"
+			GIT_TAG "vcmi-1.6"
+			BUILD_IN_SOURCE true
+			CONFIGURE_COMMAND ""
+			BUILD_COMMAND ""
+			INSTALL_COMMAND ""
+	)
```
Why is this needed for web version?

caiiiycuk replied:
I am not sure about it; it seems that vcmi-extras is something very commonly used. For example, most network games require it. It is used later to create a data file with required resources, like config. However, it is also true that the data can be bundled later and does not have to be part of the regular build. 

It is simpler to have this data built with client JS; then, you only need to provide the data from HoMM3.

IvanSavenko replied:
And a lot of players use hota. Bundle it as well? We have a lot of players from Poland. Bundle Polish translation as well?

caiiiycuk replied:
Make sense, will remove it

kambala-decapitator replied:
having such stuff in project's CML is a very bad idea. Provide a script that downloads the necessary prerequisites (that only the web versions needs) before building.

caiiiycuk replied:
Do you mean vcmi-extras or all instances of ExternalProject_Add? Why is using it bad? It's cross-platform and works well.

kambala-decapitator replied:
I'm talking about ExternalProject_Add. No other VCMI platform needs it and web version is an experimental thing atm, hence don't see a need to pollute CML with such things that are easily done with a custom script that would run before cmake invocation.

IvanSavenko commented on client/gui/UiMutex.h:
```diff
@@ -0,0 +1,16 @@
+ */
+#pragma once
+
+#ifdef VCMI_HTML5_BUILD
+typedef boost::recursive_mutex ui_mutex;
+#else
+typedef boost::mutex ui_mutex;
+#endif
```
Why different mutex is needed for web build?

caiiiycuk replied:
Good question. Probably in the first build, I didn’t have a separate thread for the server to run, and sometimes it happened that this mutex was entered multiple times from the main thread. However, in the current build, the server is executed in a separate thread, so this is probably not needed anymore. I will check this.

IvanSavenko commented on client/globalLobby/GlobalLobbyWidget.cpp:
```diff
@@ -38,7 +38,13 @@ GlobalLobbyWidget::GlobalLobbyWidget(GlobalLobbyWindow * window)
 	addCallback("closeWindow", [](int) { GH.windows().popWindows(1); });
 	addCallback("sendMessage", [this](int) { this->window->doSendChatMessage(); });
-	addCallback("createGameRoom", [this](int) { if (!CSH->inGame()) this->window->doCreateGameRoom(); });//TODO: button should be blocked instead
+	addCallback("createGameRoom", [this](int) {
+#ifndef VCMI_HTML5_BUILD
+		//TODO: button should be blocked instead
+		if (!CSH->inGame()) this->window->doCreateGameRoom();
+#endif
+	});
+
```
No wonders that we've started getting reports that lobby is broken...

dydzio0614 replied:
:rofl:

caiiiycuk replied:
Actually, I wanted to hide this button, but then I saw this TODO and thought it was okay. :)

IvanSavenko commented on server/CGameHandler.cpp:
```diff
@@ -1539,13 +1543,17 @@ void CGameHandler::save(const std::string & filename)
 			save << *this;
 		}
 		logGlobal->info("Game has been successfully saved!");
+#ifdef VCMI_HTML5_BUILD
+		html5::fsUpdate(fileName.c_str());
+#endif
```
Should be done inside `createResource` method

caiiiycuk replied:
I want to do this only for saves/settings. Is it possible to determine that from `createResource`?

IvanSavenko replied:
No, but `createResource` should not be used in other cases anyway.

IvanSavenko commented on lib/html5/stb_image.h:
Why we need stb_image as part of our sources now? Why not use sdl_image?

caiiiycuk replied:
There was a bug in Emscripten where it was not possible to use SDL_image with pthread, but it was recently fixed (https://github.com/emscripten-core/emscripten/pull/22946). I will test if this resolves the issue; if it does, then stb_image can be removed.

kambala-decapitator replied:
if really required, stb_image should be downloaded manually (submodule doesn't seem right in this case) and not be kept as copy-pasted source code

IvanSavenko commented on AI/FuzzyLite:
Why was this submodule changed?

caiiiycuk replied:
I added support for Emscripten in FuzzyLite (https://github.com/fuzzylite/fuzzylite/pull/180). Unfortunately, they merged it into the master branch only, so we need to either update FuzzyLite to the latest version or use our own fork.

IvanSavenko commented on AI/Nullkiller/CMakeLists.txt:
```diff
@@ -157,7 +157,9 @@ else()
 target_include_directories(Nullkiller PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
-target_link_libraries(Nullkiller PUBLIC vcmi fuzzylite::fuzzylite)
+if (NOT EMSCRIPTEN)
+    target_link_libraries(Nullkiller PUBLIC vcmi fuzzylite::fuzzylite TBB::tbb)
```
tbb was not here before

caiiiycuk replied:
True, it was in the previous version. I will fix it.

kambala-decapitator replied:
does NKAI no longer use tbb? As I brought this dep back in beta.

IvanSavenko reviewed (changes requested):
@caiiiycuk 
Please remove any links to pirated H3 data from your website as well as functionality to load h3 data from 3rd party website

caiiiycuk commented:
Converted to draft because we need to determine why it is running very slowly, and the requested changes may take time to implement.

caiiiycuk commented:
> @caiiiycuk
> Please remove any links to pirated H3 data from your website as well as functionality to load h3 data from 3rd party website

You are most likely referring to the vcmi-html5-launcher project. It is a launcher that, similar to vcmi native launcher, can run VCMI but only in the web version. These changes in the PR are not directly related to the launcher's code, and you can use any other launcher to run the game or even do without one altogether.

Currently, the launcher includes a link to archive.org, which I can remove and even intend to do so. However, the problem is that I am stuck between versions 1.5.7 and 1.6.3. Everything is in a disassembled state right now, and once I resolve the issues with version 1.6.3, I will remove the link to archive.org.

That said, I don’t see any reason to remove the ability to specify a link for obtaining game data. This functionality is in demand and does not violate any rules. Everything else is the responsibility of the integrator.

IvanSavenko commented:
If you're not removing links to pirated content, then I am strongly against any sort of cooperation with you. Including accepting ANY changes made by you into vcmi source code.

> This functionality is in demand and does not violate any rules

I suggest re-reading license that you have accepted when you have bought Heroes III

dydzio0614 commented:
> This functionality is in demand and does not violate any rules.

I do not care about demand for functionalities that do copyright infringement. Distributing assets of heroes 3 does that, current copyright holder aka Ubisoft has all rights to heroes 3 assets and game was not officially made free. It is also not abandonware without real copyright holder.

VCMI legality coming from being made from scratch, without distributing original game code (nor code derived from it) is one of advantages of the project that should be kept.

caiiiycuk commented:
Guys, I understand your point of view, and I agree with it. I’m not involved in distributing data from the original game or anything like that. As I already mentioned, I have a link to archive.org in the HTML5 launcher, and I will remove it. Perhaps if I re-read the license, I’ll understand why I can’t add an input field for users to paste URLs in my launcher. But I haven’t done that yet, and I don’t understand it yet. I don’t see the point in arguing; I treat you with respect and hope for the same in return. If you think it’s better not to cooperate with me, that’s fine. However, I’ll still offer my changes for review. Maybe it will help someone else.

dydzio0614 commented:
pasting URL will be mostly used to circumvent copyright by people sharing link to h3 data for others, making playing pirated game easier without having game files on PC at all

caiiiycuk commented:
@dydzio0614 Thank you for the clear explanation.

caiiiycuk commented:
Regardless of what you decide or what I decide, I just want to point out that you're overestimating the importance of the ability to insert links to resources.

1. If someone wanted to host a full game on their site, using a link insertion method wouldn't make sense—it’s an unnecessary step.  
2. Uploading resources to file-sharing services or Google Drive wouldn’t work because downloading from those sources would fail due to CORS policy restrictions.  
3. Hosting resources on their own server for a third-party site to use is economically unfeasible, as game resources typically weigh 200 MB or more. Paying for hosting so that another site can use the resources is rather odd.  

By the way, this functionality is no longer present in my launcher; I was just considering bringing it back.  

Additionally, please don’t expect me to answer for third parties or judge whether someone might use this feature for piracy. In any case, implementing this functionality isn’t difficult, even if I decide not to restore it.

dydzio0614 commented:
I will let other developers posting here decide what to do with "functionality to load h3 data from 3rd party website"

archive.org download link in launcher needs to be removed anyway

caiiiycuk commented:
> archive.org download link in launcher needs to be removed anyway

Sure! but give some time, I hope to rebuild project on this week or next week

IvanSavenko commented:
> I’m not involved in distributing data from the original game or anything like that. As I already mentioned, I have a link to archive.org in the HTML5 launcher, and I will remove it

And yet, for some magical coincidence, that archive on archive.org:
- has been uploaded when you were working on web version
- contains only minimal data files needed for vcmi and nothing else - you can't run H3 using that data

For me it looks like you're the one who uploaded it. Am I right?

caiiiycuk commented:
Yes, they appeared while I was working on the web version, but I didn’t upload anything to archive.org. If you follow your reasoning, you might be able to uncover more. By the way, regarding the claim that these links are not compatible with other versions—that's not true. They are compatible with the native VCMI, and there’s even a guide on how to use them.

```
3.2. VCMI Client (Paths for Android and Windows)

    Download the application from the developer's GitHub: https://github.com/vcmi/vcmi/releases
    (Ensure the app version matches the browser version for compatibility with browser players; currently, this is version 1.5.7.)

    Other platforms:
        Google Play for Android:
        https://play.google.com/store/apps/details?id=is.xyz.vcmi
        Flathub for Linux:
        https://flathub.org/apps/eu.vcmi.VCMI
        Ubuntu (Launchpad):
        https://launchpad.net/~vcmi/+archive/ubuntu/ppa
        macOS via Homebrew:
        brew install --cask --no-quarantine vcmi/vcmi/vcmi

Option 1) Automatic setup:

    Open the application.
    Select the language, click "Next," then press the "Copy Data" button to select the files extracted earlier from Data.zip.
    Locate the six extracted files and select them.

The app will automatically copy the files to the following locations:
For Android:

/0/Android/data/is.xyz.vcmi/files/vcmi-data/Data
For Windows:

%USERPROFILE%\Documents\My Games\vcmi\Data
```

Olegmods commented:
> Yes, they appeared while I was working on the web version, but I didn’t upload anything to archive.org. If you follow your reasoning, you might be able to uncover more. By the way, regarding the claim that these links are not compatible with other versions—that's not true. They are compatible with the native VCMI, and there’s even a guide on how to use them.

I think Ivan meant that it is for VCMI only. And it won't work with OH3 exe, or OH3-based mods like HotA or WoG/ERA.

IvanSavenko commented:
One of "economically unfeasible" websites that hosts heroes 3 data files:

```
ping homm3.zip
PING cl-msb35a6bbc.edgecdn.ru (95.181.182.182) 56(84) bytes of data.
```

```
ping sec.dos.zone
PING cl-msb35a6bbc.edgecdn.ru (95.181.182.182) 56(84) bytes of data
```

Hosted on the same IP address as yours website! These coincidences never end!

caiiiycuk commented:
> One of "economically unfeasible" websites that hosts heroes 3 data files:
> 
> ```
> ping homm3.zip
> PING cl-msb35a6bbc.edgecdn.ru (95.181.182.182) 56(84) bytes of data.
> ```
> 
> ```
> ping sec.dos.zone
> PING cl-msb35a6bbc.edgecdn.ru (95.181.182.182) 56(84) bytes of data
> ```
> 
> Hosted on the same IP address as yours website! These coincidences never end!

Ohh, you found yandex cdn.

IvanSavenko commented:
I am against accepting any changes from this person.

Can be reconsidered when any references to h3 files have been removed from his website

caiiiycuk commented:
Hi, @IvanSavenko, could you please reconsider reopening this PR? I have removed all links and the ability to pass a URL to h3 data. I just need to know if I should further refine the PR for merging into VCMI.

IvanSavenko commented:
> I have removed all links 

![зображення](https://github.com/user-attachments/assets/ca7c463e-2d87-4b0f-8a32-47041e8fde07)

Yeah, yeah, you are not involved in distributing data from the original game or anything like that, this is your evil copycat again. Right?

And I can post way more of such links if needed.

caiiiycuk commented:
>  is your evil copycat again

This thread started from this:

> Currently, the launcher includes a link to archive.org, which I can remove and even intend to do so.

I previously mentioned that I had included links to archive.org in my launcher/website. As per your request, I have removed them. Additionally, you found my Telegram message containing the same link, which I have now also removed. I have likewise removed the link from my Medium article. I would be surprised if there are any others (I might have forgotten some, but I believe there are no more links); however, I don’t mind removing all links made by me.
